go/marshal/decode.go

```diff
@@ -248,8 +250,17 @@ func sliceDecoder(t reflect.Type) decoderFunc {
 	var decoder decoderFunc
 
 	d = func(v types.Value, rv reflect.Value) {
-		slice := rv.Slice(0, 0)
-		v.(types.List).IterAll(func(v types.Value, i uint64) {
+		list, ok := v.(types.List)
+		if !ok {
+			panic(&UnmarshalTypeMismatchError{v, t, ""})
+		}
+		var slice reflect.Value
+		if rv.IsNil() {
+			slice = reflect.MakeSlice(t, 0, int(list.Len()))
```

Comment:
At some point we should probably have sanity checks that we're not trying to unmarshall huge lists. Obviously if Len doesn't fit in an int that will be a problem.